### DISCUSSION TITLE
How can I get the source code of `Image` component of NextJS ?

### DISCUSSION THREAD

**Discussion by ProPhycient5** [vercel/next.js#45510] [Help] [answered]
I want to get into the source code of the `Image` component in order to understand optimization techniques that have been implemented. As I am facing a lot of difficulties in enhancing the performance of the image on my website, that's why I need to look into it. 

Please guide or attached where `Image` component code is written.

**Comment by icyJoseph** [comment, marked as answer]
There's a two fold thing:

- Next.js may use squoosh, https://github.com/GoogleChromeLabs/squoosh during development
- Next.js tries to use sharp, https://github.com/lovell/sharp, over squoosh in production, but you must make sure sharp is installed


The code for image handling from Next.js' point of view can be found at time of writing here:

- Client side component: https://github.com/vercel/next.js/blob/canary/packages/next/src/client/image.tsx (probably not what you want)
- Server side image handling: https://github.com/vercel/next.js/blob/canary/packages/next/src/server/image-optimizer.ts, here you'll find the coupling with the two image libraries above

**Reply by ProPhycient5** [reply]
Thanks, @icyJoseph for your kind reply.  From the above links, I am actually lost to ascertain a few techniques(in code form) that have been documented on an official [site](https://nextjs.org/docs/basic-features/image-optimization) like Lazy loading, Visual Stability(by restricting CLS), Compression, Image resizing and so on. 

Actually, by taking inspiration from NextJS image optimization techniques, I want to implement them for my project(simply ReactJS project) without using NextJS.
So, can you help me ?

**Reply by icyJoseph** [reply]
>  Lazy loading

This was done with JS, but was moved to legacy: https://github.com/vercel/next.js/blob/canary/packages/next/src/client/legacy/image.tsx, there you can see usage of intersection observers to load the image

The new image uses native lazy loading, or at least favours it, which means it just uses the HTML Image lazy loading attribute: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-loading

> Visual Stability(by restricting CLS)

This is done in the client bundle, by creating a placeholder equal to the image size that'd be put in place. Search for `placeholder` in the client image source code. Granted, it is not super easy to follow statically, aka, by reading the code.

> Compression, Image resizing and so on.

This is done server side, but also a bunch of sizes and other HTML attributes are calculated client side. Look after srcSet and sizes.

All in all, it is a complex feature that has matured over 2-ish years? so the code that supports it is not precisely "easy" to find a group of lines that'll make you go AHA.

**Reply by icyJoseph** [reply]
> Actually, by taking inspiration from NextJS image optimization techniques, I want to implement them for my project(simply ReactJS project) without using NextJS.

Then you kind of want to just look at the client side images, and try to somehow generate placeholders that match up the size of the image you'd load, in order to prevent layout shifts, that can be done with CSS, and the lazy loading would just be an attribute on the HTML image element.

If you wanna have more control, use an intersection observer which would tell you whether or not the placeholder is within range of what the user is currently looking at, and if so, load the image.

**Reply by ProPhycient5** [reply]
Thanks, @icyJoseph for your elaborate reply.

Taking the reference from the above reply, I have a few doubts:
> Lazy Loading
Can we implement lazy loading by using only the `loading` attribute in `<img>`? 
or do We have to write code using IntersectionObserver API?

> Visual Stability(by restricting CLS)
How can I implement the `placeholder` attribute feature?

**Reply by icyJoseph** [reply]
Regarding lazy, it is up to you really. If you use the native loading attribute, then you depend on browser support and user adoption, as well as browser heuristics, AKA, the browser decides when it is a good time to load an image.

And intersection observer gives you more control, but at the same time, it requires main thread work, which given a large amount of images, might defeat the purpose, unless implemented in an optimal way, for instance, using only one observer and subscribing images to it, and removing them as soon as they load. 

Placeholder's that are blurred version of the image, have to be done at build time. Also the layout prop on next.js is used to calculate various CSS configuration that combined with a background image (the placeholder) ensure there's no layout shift.

**Reply by ProPhycient5** [reply]
Thanks, @icyJoseph for your lucid response.
Moreover, I explored your GitHub profile, but I didn't get any address to DM you. I expect if you could help me.

Actually last few weeks, I have stumbled video on [Good Bye useEffect](https://www.youtube.com/watch?v=HPoC-k7Rxwo) by David Khourshid and another link from beta-doc of react [You might not need useEffect](https://beta.reactjs.org/learn/you-might-not-need-an-effect), I am quite confused on which use cases, we must avoid and use `useEffect`. Because earlier most of the time, it was an easy shot for me when to use `useEffect`. So if you could recommend me any simple article which clears this confusion. Moreover, it would be great to pen an article on `useEffect`.

**Reply by icyJoseph** [reply]
Right, I am going to reserve my opinions on that talk, and cover shortly,  `useEffect` and what does React Docs say about it.

I think there's a couple that are mission critical.

## Derived state

Take for example yesterday, I was reviewing code and found a situation where:

```jsx
const Bar = ({ bar }) => {
 const [fooProps, setFooProps] = useState(bar)
 ❌❌❌
 useEffect(() => {
   setFooProps(someTransformation(bar))
 },[bar])

 return <Foo  {...fooProps} />
}
```

This is one of the most common situations where an effect+state combo was not even needed at all, because the following gets the job done:

```jsx
const Bar = ({ bar }) => {
 const fooProps = someTransformation(bar)

 return <Foo {...fooProps} />
}
```

## User Events/Actions 

Another case where effects should not be used is to act on user interaction. I once was hired to fix a bunch of bugs, one of them where two orders arrived simultaneously to an endpoint, but nobody could tell why, it happened every now and then, but it happened.

```tsx
const [product, setProduct] = useState(null)
const [orderProduct, status] = useOrderProduct()
❌❌❌
useEffect(() => {
  if (!product) return
  orderProduct(product) 
}, [product, orderProduct])

return <ProductSelector onSelect={setProduct} />
```

So you see, instead of ordering the product on user action, this anti-pattern, triggers a state update, which triggers an effect, and there the order is placed... 

In the bug I fixed, the `orderProduct` had been made, unnecessarily, into yet another hook, rather than just being a function, so when the order was triggered, a loading flag was activated, and that changed the identity of `orderProduct`, so the effect ran again. To top things, there was a race condition, because `orderProduct` was also causing a route change. Those were fun times chasing that bug.

In that particular case, some might say, the loading flag should also be used in the effect, and sure,  but eventually, as the complexity of the operation grows, so does the complexity of these useEffect's, when they could've just been a proper user event handler:

```jsx

return <ProductSelector onSelect={(product) => orderProduct(product)} />
```

## Data fetching

 With data fetching, because as the application grows, you encounter a few things. Not an exhaustive list: 

- Same server data is needed in various places
- Server data needs to be mutated and these updates should be visible in the client 
- Communication with the server goes through various stages, idle, pending, resolved, rejected

This inevitably results in a mini-framework within your application. This is not just for React apps, but any app that consumes server data heavily. So, for production apps, instead of dividing your attention between business code and your mini-framework, you should just use a data-fetching/global management solution, such as SWR, React-Query, Zustand, Jotai, etc...

---

I think perhaps a good rule of thumb is, if some rendering pass should notify _others_ about state changes, then an effect is a good option. The render pass itself should only generated React Elements, aka, be pure. 

I think most examples in https://beta.reactjs.org/learn/you-might-not-need-an-effect are pretty straight forward. 

They do open the section by mentioning:

- You don’t need Effects to transform data for rendering.
- You don’t need Effects to handle user events.

And the examples shown are in one way or another, related to these two. 

I think some people find this one a bit strange, https://beta.reactjs.org/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes - change the key to reset state, because a key change tells React that there's a totally new child in place, so state must not be shared.

Perhaps the most "new" is that if some external service needs to inform React about a stateful changes, `useSyncExternalStore` is the tool.

**Reply by ProPhycient5** [reply]
Thanks, @icyJoseph for your elaborate response on a vital topic.
